corehq/db.py

```diff
@@ -1,13 +1,82 @@
 from django.conf import settings
-from sqlalchemy.engine import create_engine
+import sqlalchemy
 from sqlalchemy.orm.scoping import scoped_session
 from sqlalchemy.orm.session import sessionmaker
 from django.core import signals
 
-_engine = create_engine(settings.SQL_REPORTING_DATABASE_URL)
-_session_factory = sessionmaker(bind=_engine)
+DEFAULT_ENGINE_ID = 'default'
 
-Session = scoped_session(_session_factory)
+
+def create_engine(connection_string=None):
+    # todo: this function is just a proxy for the sqlalchemy version and should be removed
+    connection_string = connection_string or settings.SQL_REPORTING_DATABASE_URL
+    return sqlalchemy.create_engine(connection_string)
+
+
+class SessionFactory(object):
+    """
+    Shim class for a single connection/session factory
+    """
+
+    def __init__(self, connection_string):
+        self.engine = create_engine(connection_string)
+        self._session_factory = sessionmaker(bind=self.engine)
+        # Session is the actual constructor object
+        self.Session = scoped_session(self._session_factory)
+
+
+class ConnectionManager(object):
+    """
+    Object for dealing with sqlalchemy engines and sessions.
+    """
+
+    def __init__(self):
+        self._session_factories = {}
+
+    def _get_or_create_factory(self, engine_id):
+        if engine_id not in self._session_factories:
+            self._session_factories[engine_id] = SessionFactory(self._get_connection_string(engine_id))
+        return self._session_factories[engine_id]
+
+    def get_session_constructor(self, engine_id=DEFAULT_ENGINE_ID):
+        """
+        This returns a class that can be instantiated and will have a session scoped
+        to the local thread.
+        """
+        return self._get_or_create_factory(engine_id).Session
+
+    def get_engine(self, engine_id=DEFAULT_ENGINE_ID):
+        """
+        Get an engine by ID. This should be a unique field identifying the connection,
+        e.g. "report-db-1"
+        """
+        return self._get_or_create_factory(engine_id).engine
+
+    def dispose_engine(self, engine_id=DEFAULT_ENGINE_ID):
+        """
+        If found, disposes an engine and removes it from this, otherwise does nothing.
+        """
+        if engine_id in self._session_factories:
+            factory = self._session_factories.pop(engine_id)
```

Comment:
Yes, we should definitely call Session.remove() first. Extract form the docs:

"Connections that are checked out are not discarded when the engine is disposed or garbage collected, as these connections are still strongly referenced elsewhere by the application. However, after Engine.dispose() is called, those connections are no longer associated with that Engine; when they are closed, they will be returned to their now-orphaned connection pool which will ultimately be garbage collected, once all connections which refer to it are also no longer referenced anywhere. Since this process is not easy to control, it is strongly recommended that Engine.dispose() is called only after all checked out connections are checked in or otherwise de-associated from their pool." 
http://docs.sqlalchemy.org/en/rel_0_9/core/connections.html#engine-disposal